web/packages/teleterm/src/services/config/configService.ts

```diff
@@ -168,14 +200,28 @@ function getDefaultTerminalFont(platform: Platform) {
   }
 }
 
-export function createConfigService(
-  appConfigFileStorage: FileStorage,
-  platform: Platform
-) {
-  return createConfigStore(
-    createAppConfigSchema(platform),
-    appConfigFileStorage
-  );
+function getKeyboardShortcutDescription(about: string): string {
+  return `Shortcut to ${about}. A valid shortcut contains at least one modifier and a single key code, for example "Ctrl+Shift+A". Function keys do not require a modifier.`;
+}
+
+export function createConfigService({
+  configFile,
+  configJsonSchemaFile,
+  platform,
+}: {
+  configFile: FileStorage;
+  configJsonSchemaFile: FileStorage;
+  platform: Platform;
+}) {
+  const schema = createAppConfigSchema(platform);
+
+  updateJsonSchema({
+    configSchema: schema,
+    configFile,
+    configJsonSchemaFile,
+  });
+
+  return createConfigStore(schema, configFile);
```

Comment:
This is not a part of this PR so don't feel pressured to change it, but I think this function could be refactored. I'm inside `createConfigService`, what is a config store? What's more, to understand what config service does I need to understand `createConfigStore` first. To do that, I have to jump to another file.

IMHO it could be all put into the same file. There's a single test for `updateJsonSchema` which could be expressed as a test for `createConfigService` really.

Reply:
So the idea behind `createConfigService`/`createConfigStore` is that it allows you for keeping the service implementation and the app config schema separately. It improves readability and makes testing easier - I can use a small schema for testing purposes.

Now I feel like I have broken this rule by putting `updateJsonSchema` in `createConfigService` (it should be rather put in `createConfigStore` to keep the implementation in one place).

Probably we could extract a function `createAppConfigSchema` to the separate file and move `createConfigStore` to `createConfigService`. However, we still have to call `createAppConfigSchema` somewhere. I think it is better to do it in `configService` module rather than in `main.ts`. What do you think?

Reply:
Yes, this my main motivation. I also wouldn't want to have a "low level" service implementation (which currently sits in `createAppStore` in the same file as the "high-level" app config.

Reply:
In terms of testing, IMHO a better choice would be to provide the real schema as the default argument to `createConfigService` (`schema = createAppConfigSchema()`). Mocking stuff by ripping out a whole file import should be a last resort. The problem is that to create a schema we need to know the platform so it's not as straightforward here. 

The way I see it is that `createConfigService` needs a schema, it doesn't really care how the schema is created. But the way it's written right now means that how the schema is created is hardcoded into `createConfigService`.

It's the functional equivalent of hardcoding class initialization in OOP, something like:

```typescript
class ConfigService {
  constructor(platform) {
    this.schema = new ConfigSchema(platform)
  }
}

// instead of

class ConfigService {
  constructor(private schema) {
  }
}
```

But zooming out even more, I'd consider what do we gain by using a smaller schema in tests! It's not like it ensures that `createConfigStore` works with any kind of schema. In our app there's a single schema that we need to support.

---

> I also wouldn't want to have a "low level" service implementation (which currently sits in `createAppStore` in the same file as the "high-level" app config.

I agree here though not because of the high/low level distinction but rather because the creation of schema is irrelevant to how the config works.

---

I think we've arrived at two possible solutions here:

* Keep schema creation hardcoded into the service, get rid of special schema in tests and just test the whole thing.
* Separate  schema creation from the service creation so that `services/config` exposes `createConfigService` and `createConfigSchema`, then create schema in main.ts first followed by the call to `createConfigService`.

I suspect that the second option might be more up your alley. 😏

As I said before, the biggest problem for me is that `createConfigService` is supposed to return I guess a service of some sorts that you operate on. But inbetween it introduces a concept of a store that is what you actually end up using. So it looks like splitting `createConfigService` implementation between across schema creation & config service API would solve this problem. Creating the schema would be defined in another file while the config service implementation lives in another.

Reply:
>Separate schema creation from the service creation so that services/config exposes createConfigService and createConfigSchema, then create schema in main.ts first followed by the call to createConfigService.

I realized that this would create a small inconvenience. In `createConfigService` we also export `ConfigService` type which is an inferred type: `ReturnType<typeof createConfigService>`. So, if would like to move `createConfigSchema` to `main` then this inferred type can't work. 
It seems to me that `services/config` should expose a ready-to-use function, so the caller doesn't have to think about proving the schema implementation.

Now, I'm more in favour:
>Keep schema creation hardcoded into the service, get rid of special schema in tests and just test the whole thing.

But the question is: how to test `updateJsonSchema` output? For my small schema, the output was a few lines, for our real app schema it is over 100 lines which I feel weird to copy to the test file :D

Reply:
In terms of the inferred type, I assume that if schema-handling would be defined in a separate file, then that file would do `export type AppConfig = z.infer<ReturnType<typeof createAppConfigSchema>>`.

Couldn't the equivalent of `createConfigStore` work on a specific schema and shape types instead of on a generic one like it does now?

If it can work on specific types then we could get rid of `ReturnType<typeof createConfigService>`, couldn't we?

---

> But the question is: how to test `updateJsonSchema` output? For my small schema, the output was a few lines, for our real app schema it is over 100 lines which I feel weird to copy to the test file :D

We could just check if `  expect(configFile.get('$schema')).toBe('config_schema.json')` and that `replace` was called on `jsonSchemaFile` and that's it.

Comparing the output of `zodToJsonSchema` doesn't bring a lot of value IMHO – in the end it's almost like a snapshot test of zod-to-json-schema and we typically don't write test for our dependencies.

Reply:
> Couldn't the equivalent of createConfigStore work on a specific schema and shape types instead of on a generic one like it does now?

I merged `createConfigStore` and `createConfigService` into one, having a generic _config store_ was a bit of overengineering. As you said, we have only one type of config, so it wasn't really useful.

>We could just check if   expect(configFile.get('$schema')).toBe('config_schema.json') and that replace was called on jsonSchemaFile and that's it.

Yeah, that should be fine.